fix(core): Correctly track usage

abhiaiyer91 opened the pull request:
- Fix usage reporting
- Fix usage tracking

## Description

<!-- Provide a brief description of the changes in this PR -->

## Related Issue(s)

<!-- Link to the issue(s) this PR addresses, using hashtag notation: #123 -->

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Stream usage now reports the most recent step's tokens, fixing incorrect cumulative tracking.

* **Improvements**
  * Provider metadata (providerMetadata/providerOptions) is propagated and preserved more consistently across message parts, UI/DB conversions, and tool-call payloads.
  * Usage objects expose additional optional token fields (cachedInputTokens, reasoningTokens).

* **Tests**
  * Added and expanded usage-tracking and provider-metadata preservation tests; adjusted snapshots and expectations.

* **Chores**
  * Added a patch changeset and bumped several dev dependency versions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

abhiaiyer91 committed:
Fix usage reporting

abhiaiyer91 committed:
Fix usage tracking

abhiaiyer91 committed:
changeset

CalebBarnes commented on packages/core/src/agent/message-list/index.ts:
```diff
@@ -1497,6 +1497,13 @@ export class MessageList {
+            const toolCallProviderMetadata = {
+              ...('providerOptions' in coreMessage && coreMessage.providerOptions ? coreMessage.providerOptions : {}),
+              ...('providerOptions' in part && part.providerOptions ? part.providerOptions : {}),
+            };
```
@TylerBarnes I think you had some thoughts on adding the providerOptions here. Maybe we only want to include the providerOptions on the individual parts instead of also adding from the top level message?

LekoArts committed:
Update eager-bags-drive.md

CalebBarnes committed:
Merge branch 'main' into usage-fix

CalebBarnes committed:
fix message providerMetadata (#10227)

## Description

<!-- Provide a brief description of the changes in this PR -->

## Related Issue(s)

<!-- Link to the issue(s) this PR addresses, using hashtag notation:
#123 -->

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if
applicable)
- [ ] I have added tests that prove my fix is effective or that my
feature works

---------

Co-authored-by: CalebBarnes <CalebBarnes@users.noreply.github.com>

CalebBarnes committed:
dont test for merged providerOptions

CalebBarnes committed:
remove duplicate test

CalebBarnes committed:
add todos to providerOptions testing conversions using unsupported UI message

CalebBarnes committed:
add todos to tests expecting message-level providerOptions to be preserved through db -> ui -> model message conversions

CalebBarnes committed:
update snapshots to remove providerMetadata: undefined

CalebBarnes committed:
update result.usage snapshot with undefined values
